Framework/Core/src/DataAllocator.cxx

```diff
@@ -177,6 +177,16 @@ void DataAllocator::adopt(const Output& spec, TableBuilder* tb)
   context->addBuffer(std::move(header), buffer, std::move(finalizer), channel);
 }
 
+void DataAllocator::snapshot(const Output& spec, const char* payload, size_t payloadSize,
+                             o2::header::SerializationMethod serializationMethod)
+{
+  auto proxy = mContextRegistry->get<MessageContext>()->proxy();
+  FairMQMessagePtr payloadMessage(proxy.createMessage(payloadSize));
+  memcpy(payloadMessage->GetData(), payload, payloadSize);
```

Comment:
I think there is no need to memcpy here - if you have an already serialized payload (as in ready for shipping) it's buffer should be under FairMQ control and the "snapshotting" should be delegated to fairmq which can do this without copying (in principle and I think actually in most cases).

Reply:
Don't you have to move the ownership if you do that, though? This would cause the `unique_ptr` to  the incoming message to become a `nullptr` which breaks the current semantic of  `snapshot`.

This also would have implications when using shared memory with the current ownership model because it's not clear anymore who owns the message in case there are multiple subscribers to it. Fixing this latter issue would require we move to the unmanaged region model, which albeit on the todo list is not yet done.

Reply:
no need to move anything, just request a message copy, you get a new new unique ptr and don't even know that the underlying resource is refcounted (@rbx ).
For unmanaged regions in the cases where the payload is not managed by a fairmqmessage this is of course all manual.
Anyway - if the API/semantics force the implementation to do anything, especially copy data around, the API needs to be redesigned.

What problem are you trying to solve exactly with this new addition?

Reply:
Requesting a shallow message copy still breaks `snapshot` semantics which requires the snapshot and the source of the snapshot be independent the one from the other. I think this is a perfectly valid use case to cover, e.g. for histograms or for building incrementally multiple messages without having to recreate them from scratch every time. This is merely a tradeoff between a memory copy and object initialisation.

That said, given the idea here (@knopers8 correct if I am wrong) is to reroute messages without having to deserialise and serialise them again, maybe we need to have a "reroute" method. So what I would suggest is that we keep this as is and we add reroute to the wishlist...

@matthiasrichter any opinions?

Reply:
To get a copy you can do:
```C++
FairMQMessagePtr msg(NewMessage(size));
FairMQMessagePtr msgCopy(NewMessage());
msgCopy->Copy(*msg);
```
Depending on the transport this will be done without actual copy of the data. ZeroMQ - no copy (ref counted), shmem - with copy (implementation does memcpy). The ZeroMQ implementation historically relies on a method provided by zmq itself which does the ref counting. We should probably also implement it via hard copy to keep it consistent (a message copy should be safe to modify after the copy takes place).

Reply:
in o2 the idea was a message (a buffer) once constructed is immutable, so removing the possibility of ref counted zeromq message copies would be a bummer... we do have a number of cases where the same data needs to be sent more than once, say for QC or monitoring. But anyway, this is something that should sit in fairmq and be (or not) efficient, no need for us (o2) to reason about it.

As for the snapshot - if the user wants to reuse some data if initialization is expensive the more power to them, but that should happen in user code. Exposing an API that advertises buffer mutability at the framework level is in my opinion a mistake and it definitely breaks the basic data model assumptions.

Reply:
`snapshot` is for buffers / objects which are (in general) coming from outside the framework, so your statement holds: once snapshotted, copied data (which is owned by the framework) is immutable.

Reply:
glad to hear that. in that case i'm all ok with the memcpy. i'll just buy some more cpu and memory from my allowance.

Reply:
I had a look around DPL to see what would need to be done to copy the incoming **message**, not **its contents** which are available in the `InputRecord` API (pointers to FairMQMessages are not).
I guess that the message pointers would need to be exposed in the `DataRef` (as `const FairMQMessage*`) among the existing fields. To do that, `InputSpan` should return message pointers instead of `ptr->GetData()` (though it seems like its whole purpose was to shield the further code from FairMQ). The getters in `InputRecord` should be modified accordingly to return the updated `DataRef` structure.
Then we would be able to pass the message pointer to the some DataAllocator's method, which can actually do:
```
FairMQMessagePtr msgCopy(NewMessage());
msgCopy->Copy(inputDataRef->message);
```
As I understand, we don't have to worry about the unique_ptr destroying the message, because inside it is refcounted.

Is that the way forward that everyone is satisfied with? I can try to give it a go, though it seems to break some initial design decisions.

Reply:
I would prefer if the InputSpan does not know anything about the actual transport, so that we can still attack the "run algorithms without fairmq" part which has been requested by multiple parties (including you). How about exposing a different service (i.e. not a InputRecord) which allows you do the rerouting?

Reply:
It's good with me